src/librustc_middle/ty/sty.rs

```diff
@@ -2371,27 +2398,25 @@ impl<'tcx> Const<'tcx> {
                 // and we use the original type, so nothing from `substs`
                 // (which may be identity substs, see above),
                 // can leak through `val` into the const we return.
-                Ok(val) => Const::from_value(tcx, val, self.ty),
-                Err(ErrorHandled::TooGeneric | ErrorHandled::Linted) => self,
-                Err(ErrorHandled::Reported(ErrorReported)) => tcx.const_error(self.ty),
+                Ok(val) => Some(Ok(val)),
+                Err(ErrorHandled::TooGeneric | ErrorHandled::Linted) => None,
+                Err(ErrorHandled::Reported(e)) => Some(Err(e)),
             }
         } else {
-            self
+            None
         }
     }
+}
 
+impl<'tcx> Const<'tcx> {
     #[inline]
     pub fn try_eval_bool(&self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> Option<bool> {
-        self.try_eval_bits(tcx, param_env, tcx.types.bool).and_then(|v| match v {
-            0 => Some(false),
-            1 => Some(true),
-            _ => None,
-        })
+        self.val.eval(tcx, param_env).try_to_bool()
     }
 
     #[inline]
     pub fn try_eval_usize(&self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> Option<u64> {
-        self.try_eval_bits(tcx, param_env, tcx.types.usize).map(|v| v as u64)
+        self.val.eval(tcx, param_env).try_to_machine_usize(tcx)
```

Comment:
reduces interning

Reply:
This also reduces `layout_of` invocations, because instead of invoking that on `tcx.types.usize`, we just use the size in `try_to_machine_usize`